finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala

```diff
@@ -0,0 +1,182 @@
+    val bytes = "%s:%s".format(c.username, c.password).getBytes(Charsets.Utf8)
+    "Basic " + Base64StringEncoder.encode(bytes)
+  }
+
+  private[this] def fail(ctx: ChannelHandlerContext, t: SocketAddress => Throwable): Unit = {
+    // We "try" because it might be already cancelled and we don't need to handle
+    // cancellations here - it's already done by `proxyCancellationsTo`.
+    connectFuture.get().setFailure(t(ctx.getChannel.getRemoteAddress))
+
+  }
+
+  private[this] def fail(c: Channel, t: SocketAddress => Throwable) {
+    Option(connectFuture.get) foreach { _.setFailure(t(c.getRemoteAddress)) }
+    Channels.close(c)
+  }
+
+  override def connectRequested(ctx: ChannelHandlerContext, e: ChannelStateEvent): Unit = {
+    println("entering connectRequested")
+    e match {
+      case de: DownstreamChannelStateEvent =>
+        if (!connectFuture.compareAndSet(null, e.getFuture)) {
+          fail(ctx.getChannel, (remoteAddress : SocketAddress) => new InconsistentStateException(remoteAddress))
+          return
+        }
+
+        // proxy cancellation
+        val wrappedConnectFuture = Channels.future(de.getChannel, true)
+        de.getFuture.addListener(new ChannelFutureListener {
+          def operationComplete(f: ChannelFuture) {
+            if (f.isCancelled)
+              wrappedConnectFuture.cancel()
+          }
+        })
+
+        // Proxy failures here so that if the connect fails, it is
+        // propagated to the listener, not just on the channel.
+        wrappedConnectFuture.addListener(new ChannelFutureListener {
+          def operationComplete(f: ChannelFuture) {
+            if (f.isSuccess || f.isCancelled) {
+              return
+            }
+
+            fail(f.getChannel, (_: SocketAddress) => f.getCause)
+          }
+        })
+
+        // We propagate the pipeline with a new promise thereby delaying the original connect's
+        // satisfaction.
+        // TODO: Figure out why the remoteAddress is
+        val wrappedEvent = new DownstreamChannelStateEvent(
+          de.getChannel, wrappedConnectFuture,
+          de.getState, de.getValue)
+
+
+        println(s"Connecting to ${wrappedEvent.getChannel.getRemoteAddress} with $wrappedEvent")
+        super.connectRequested(ctx, wrappedEvent)
+    }
+
+  }
+
+  override def writeRequested(ctx: ChannelHandlerContext, e: MessageEvent): Unit = {
+    println(s"Storing $e for now")
+    super.writeRequested(ctx, e)
+  }
+
+  override def channelConnected(ctx: ChannelHandlerContext, e: ChannelStateEvent): Unit = {
+    println("channelConnected")
+    if (connectFuture.get eq null) {
+      fail(ctx, (remoteAddress: SocketAddress) => new InconsistentStateException(remoteAddress))
+      return
+    }
+    // Add HTTP client codec so we can talk to an HTTP proxy.
+    ctx.getPipeline().addBefore(ctx.getName(), httpCodecKey, httpClientCodec)
+
+    val httpProxyConnectRequestFuture = Channels.future(e.getChannel, true)
```

Comment:
You don't need it. Just use `connectFuture` (is what Finagle transport waiting on).